nixos/modules/services/misc/languagetool.nix

```
@@ -0,0 +1,71 @@
+{ config, lib, options, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.languagetool;
+  settingsFormat = pkgs.formats.javaProperties {};
+in {
+  options.services.languagetool = {
+    enable = mkEnableOption "the LanguageTool server";
+
+    port = mkOption {
+      type = types.port;
+      default = 8081;
+      example = 8081;
+      description = mdDoc ''
+        Port on which LanguageTool listens.
+      '';
+    };
+
+    public = mkEnableOption "access from anywhere (rather than just localhost)";
+
+    allowOrigin = mkOption {
+      type = types.nullOr types.str;
+      default = null;
+      example = "https://my-website.org";
+      description = mdDoc ''
+        Set the Access-Control-Allow-Origin header in the HTTP response,
+        used for direct (non-proxy) JavaScript-based access from browsers.
+        `null` to allow access from all sites.
+      '';
+    };
+
+    settings = lib.mkOption {
+      type = settingsFormat.type;
+      default = {
+        cacheSize = mkDefault 1000;
+      };
```

```suggestion
      type = types.submodule {
        freeformType = settingsFormat.type;
        
        options.cacheSize = mkOption ...
      };
      default = {};
```
set defaults configured like this are overwritten completely if any key in the set is written to. better to use a freeform submodule to avoid that.

---

Interesting. Modules are more intricate than they appear at first glance :-)

If I set
```
      type = types.submodule {
        freeformType = settingsFormat.type;
        
        options.cacheSize = mkOption {
          type = types.ints.unsigned;
          default = 1000;
          description = mdDoc "Number of sentences cached.";
        };
      };
```
it results in the error `jq: error (at /build/.attr-068668c42wmzphgng845sgchm055ia1dqjnakbxcqmdda96l0hnd:0): number (1000) cannot be matched, as it is not a string` because the `javaProperties` format returns an `attrsOf str`.

If I set the type to `types.str;` it works, but it's a little odd. Is there a nice way to cast numeric options to strings in this context?

---

you could set `apply = toString` in the option definition

---

that's very nice!